compiler/generator.go

```diff
@@ -175,20 +173,29 @@ func (g *generator) expr(schema *jsonschema.Schema) (ast.Expr, []*ast.ImportSpec
 	}
 
 	// Handle object types that are emitted as Go map types (not named struct types).
-	if len(schema.Type) == 1 && schema.Type[0] == jsonschema.ObjectType && schema.Properties == nil && schema.AdditionalProperties != nil {
+	if isTypeOrNull(schema, jsonschema.ObjectType) && schema.Properties == nil && schema.AdditionalProperties != nil {
 		typeExpr, imports, err := g.expr(schema.AdditionalProperties)
 		if err != nil {
 			return nil, nil, err
 		}
 		return &ast.MapType{Key: ast.NewIdent("string"), Value: typeExpr}, imports, nil
 	}
 
+	nullable := isNullable(schema)
 	// Handle types represented by Go builtin types or some other non-named types.
-	if len(schema.Type) != 1 && (schema.Go == nil || !schema.Go.TaggedUnionType) {
+	if (nullable && len(schema.Type) != 2 || !nullable && len(schema.Type) != 1) && (schema.Go == nil || !schema.Go.TaggedUnionType) {
 		return emptyInterfaceType, nil, nil
 	}
-	if len(schema.Type) == 1 && goBuiltinType(schema.Type[0]) != "" {
-		return ast.NewIdent(goBuiltinType(schema.Type[0])), nil, nil
+	if nullable && len(schema.Type) == 2 || !nullable && len(schema.Type) == 1 {
```

Comment:
can you add some paranthesis in here? I can't remember if `&&` is evalled before `||`, so not sure if this is correct.